Userland/Libraries/LibCore/SharedCircularQueue.h

```diff
@@ -0,0 +1,239 @@
+        while (true) {
+            // (3) The >= is not strictly necessary, but it feels safer :^)
+            if (next >= m_queue->m_queue->m_tail.load())
+                return QueueStatus::Empty;
+
+            // (4) This CAS only succeeds if nobody modified the head while we did the above work.
+            if (m_queue->m_queue->m_head.compare_exchange_strong(next, next + 1))
+                // (5)
+                return m_queue->m_queue->m_data[next % Size];
+        }
+    }
+
+private:
+    struct SharedMemorySPCQ {
+        explicit SharedMemorySPCQ() = default;
+        SharedMemorySPCQ(SharedMemorySPCQ const&) = delete;
+        SharedMemorySPCQ(SharedMemorySPCQ&&) = delete;
+        ~SharedMemorySPCQ() = default;
+
+        // Invariant: tail >= head
+        // Invariant: head and tail are monotonically increasing
+        // Invariant: tail always points to the next free location where an enqueue can happen.
+        // Invariant: head always points to the element to be dequeued next.
+        // Invariant: tail is only modified by enqueue functions.
+        // Invariant: head is only modified by dequeue functions.
+        // An empty queue is signalled with:  tail = head
+        // A full queue is signalled with:  head - 1 mod size = tail mod size  (i.e. head and tail point to the same index in the data array)
+        // FIXME: These invariants aren't proven to be correct after each successful completion of each operation where it is relevant.
+        //        The work could be put in but for now I think the algorithmic correctness proofs of the functions are enough.
+        Atomic<size_t, AK::MemoryOrder::memory_order_seq_cst> m_tail { 0 };
+        Atomic<size_t, AK::MemoryOrder::memory_order_seq_cst> m_head { 0 };
+
+        alignas(ValueType) Array<ValueType, Size> m_data;
+    };
+
+    class RefCountedSharedMemorySPCQ : public RefCounted<RefCountedSharedMemorySPCQ> {
+        friend class SharedSingleProducerCircularQueue;
+    public:
+        SharedMemorySPCQ* m_queue;
+        void* m_raw;
+        int m_fd;
+
+        ~RefCountedSharedMemorySPCQ()
+        {
+            MUST(System::close(m_fd));
+            MUST(System::munmap(m_raw, sizeof(SharedMemorySPCQ)));
+            dbgln("destructed SSPCQ at {:p}, shared mem: {:p}", this, this->m_raw);
+        }
+
+    private:
+        RefCountedSharedMemorySPCQ(SharedMemorySPCQ* queue, int fd) 
+            : m_queue(queue)
+            , m_raw(reinterpret_cast<void*>(queue))
+            , m_fd(fd)
+        {
+        }
+
+    };
+
+    static ErrorOr<SharedSingleProducerCircularQueue<T, Size>> try_create_internal(int fd, bool is_new)
+    {
+        auto name = String::formatted("SharedSingleProducerCircularQueue@{:x}", fd);
+        auto* raw_mapping = TRY(System::mmap(nullptr, sizeof(SharedMemorySPCQ), PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0, 0, name));
+        dbgln("successfully mmapped {} at {:p}", name, raw_mapping);
```

Comment:
These `dbgln(..)`'s should be under a macro?

Reply:
They could be, but they're not very spammy so I think it's fine to leave them.